### PR TITLE
[WIP] check disk size before consuming disk space

### PR DESCRIPTION
**Pull request by eriknordmark**
Since we have time, did this complete fix where we are aware of potential changes in disk usage when we purge.

Note that the two commits are unrelated, but I don't have sufficient test bandwith and devices to test from too many different branches.

### CODE REVIEW

**Comment by kalyan-nidumolu** [comment]
I did not understand the fix or the issue.. Can u describe the issue and the workaround?

**Comment by eriknordmark** [comment]
> I did not understand the fix or the issue.. Can u describe the issue and the workaround?

@kalyan-nidumolu issue was described and discussed on slack.

The fundamental issue is that this check is done after the disk space is already consumed, hence returning an error and failing the application doesn't prevent the disk space from being used. This workaround fix does not address that.
Then with volumemgr and purge (using the same image) thing changed a bit because the new volume is created in volumemgr (as opposed to hidden in domainmgr in the old code).
This workaround addresses the latter by not having his check in the purge path - with that approach the disk space used by the pre-purge volume will be freed up (as opposed to consumed until the user deletes the app instance).

But we should fix the fundamental issue and move the check to doInstall before the volumes are created and check their maxSizeBytes.

**Comment by eriknordmark** [comment]
> Makes sense as a stopgap @eriknordmark -- but as you said -- this really needs to be addressed as part of the volume management refactoring properly.

Since this isn't urgent let me move the check up front as part of this PR.

File: pkg/pillar/cmd/zedmanager/updatestatus.go [hunk 1]
```diff
@@ -206,7 +206,7 @@ func checkDiskSize(ctxPtr *zedmanagerContext) error {
 		if err != nil {
 			log.Errorf("checkDiskSize: err: %s", err.Error())
 			// Assume application is going down and its disks
-			// are going away.
+			// are going away. XXX which is utterly incorrect
```
Comment (kalyan-nidumolu) [inline pkg/pillar/cmd/zedmanager/updatestatus.go:209]: GetDiskSizeForAppInstance  can fail when error when "diskmetrics.GetDiskVirtualSize" fails..

Its just better to remove the comment here.. No point in retaining an incorrect comment.

There is a bigger question about accounting:

1) GetDiskSizeForAppInstance --> This is skipping RO images. Need to take that into account
        --> Should we count them separately - as separate from space take for App Instance?
        --> If the same AppImage is used for multiple App Instances - we definitely should not be double counting - hence the reason for skipping the check.
       --> This needs to be adjusted for the CAS world too - some layers can be shared across multiple applications

2) GetDiskSizeForAppInstance --> This is actually looking at the downloaded file and getting the Virtual size out of it - So the check will not actually work if the file is not downloaded.

File: pkg/pillar/zedcloud/send.go [hunk 2]
```diff
@@ -211,15 +211,15 @@ func VerifyAllIntf(ctx *ZedCloudContext,
...
+			url, errorList)
 		log.Errorln(errStr)
 		return false, remoteTemporaryFailure, intfStatusMap, errors.New(errStr)
 	}
 	if intfSuccessCount < successCount {
 		errStr := fmt.Sprintf("Not enough Ports (%d) against required count %d"+
-			" to reach Zedcloud; last failed with %v. intfStatusMap: %+v",
-			intfSuccessCount, successCount, errorList, intfStatusMap)
+			" to reach Zedcloud; last failed with %v",
+			intfSuccessCount, successCount, errorList)
```
Comment (kalyan-nidumolu) [inline pkg/pillar/zedcloud/send.go:222]: Unrelated changes? :-)

Comment (eriknordmark) [reply pkg/pillar/zedcloud/send.go:222]: No, that is a distinct commit in the same PR. (I guess I should have said the commits should be reviewed independently)

File: pkg/pillar/cmd/zedmanager/updatestatus.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedmanager/updatestatus.go]: Checking the Size of the Drive we get from the controller.

**Comment by eriknordmark** [comment]
> LGTM, but I guess this needs to be reconciled (eventually) with what @adarsh-zededa and @deitch are cooking for CAS/Images/Snapshots

Agreed. And as part of that we might get a better handle on exactly how much disk space we use both while creating a volume and also when it is created, so we can do better checks.
